### PR TITLE
Basic OSS framework for running benchmarks on Delta workloads

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

Add a basic framework for writing benchmarks on Spark running in a EMR cluster. Here is how to use it. 
- Start your own EMR cluster, and get the hostname and the pem file to ssh into it.
- The file `delta/benchmarks/run-benchmark.py` has a list of benchmarks as a map between benchmark name --> specification (explained further below).
- In the directory `delta/benchmarks/`, run a selected benchmark as  `./run-benchmark.py --cluster-hostname <hostname> -i <pem file> --benchmark <benchmark-name> `. 
- This will eventually produce an output like this. 
```
...
There is a screen on:
	18319..ip-172-31-54-89	(Detached)

Files for this benchmark:
20220107-131623-test-benchmarks.jar
20220107-131623-test-cmd.sh
20220107-131623-test-out.txt
>>> Benchmark script started in a screen. Stdout piped into 20220107-131623-test-out.txt.Final report will be generated on completion in 20220107-131623-test-report.json.
```
- You can then ssh into the EMR cluster to monitor it.

The actual spark code of a benchmark is defined in the scala files that are compiled into a jar using build.sbt. The benchmark specifications in `delta/benchmarks/run-benchmark.py` defines the following 
- which main scala class to be started.
- command line argument for the main function
- additional maven artifact to load (example `io.delta:delta:1.0`)
- spark confs

The script `run-benchmark.py` does the following:
- compile the scala code into a fat jar
- upload it to the given hostname
- using ssh to the hostname, it will launch a screen and start the main class with spark-shell/spark-submit
- Monitor the output file with continuous logging and wait for the completion of the workload
- Upload the generated benchmarks results as csv/json to the given 

Structure of the code
- `build.sbt`, `project/`, `src/` form the SBT project.
- `Benchmark.scala` is the basic interface, and `TestBenchmark.scala` is a test implementation.
- `scripts` has the core python scripts that are called by `run-benchmark.py`

## How was this patch tested?
Manual testing using the following command:

`./run-benchmark.py --cluster-hostname <hostname<  -i <pem-file>  --benchmark-path <s3-path> --benchmark test`

### CODE REVIEW

File: benchmarks/scripts/benchmarks.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+                                           throw_on_error=False)
+            if completed_file in out.decode("utf-8"):
+                completed = True
+            else:
+                time.sleep(60)
+
+        # Check the last few lines of output files to identify success
+        (_, out, _) = run_cmd_over_ssh(f"tail {output_file}", cluster_hostname, ssh_id_file,
+                                       throw_on_error=False)
+        if "SUCCESS" in out.decode("utf-8"):
```
Comment (allisonport-db) [inline benchmarks/scripts/benchmarks.py:269]: The JSON report is printed after `SUCCESS` so successful runs are being reported as a failure here.

File: benchmarks/README.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+#### Run simple test workload
+Verify that you have the following information
+  - <HOST_NAME>: EMR cluster master node host name
+  - <PEM_FILE>: Local path to your PEM file for SSH into the master node.
+  - <BENCHMARK_PATH>: S3 path where tables will be created. Make sure your AWS credentials have read/write permission to that path.
+    
+Then run a simple table write-read test: Run the following in your shell. 
+    ```sh
+    ./run-benchmark.py --cluster-hostname <HOSTNAME> -i <PEM_FILE> --benchmark-path <BENCHMARK_PATH> --benchmark test
```
Comment (allisonport-db) [inline benchmarks/README.md:37]: This is running with python 2 on my machine. I had to use `python3 run-benchmark.py --cluster-hostname ...`

File: benchmarks/README.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    ```
+    The above metrics are also written to a json file and uploaded to the given path. Please verify that both the table and report are generated in that path. 
+
+
+## Internals of the framework
+
+Structure of this framework's code
+- `build.sbt`, `project/`, `src/` form the SBT project which contains the Scala code that define the benchmark workload.
+    - `Benchmark.scala` is the basic interface, and `TestBenchmark.scala` is a sample implementation.
+- `run-benchmark.py` contains the specification of the benchmarks defined by name (e.g. `tpcds-3tb-delta`). Each benchmark specifications is defined by the following:
```
Comment (allisonport-db) [inline benchmarks/README.md:103]:
```suggestion
- `run-benchmark.py` contains the specification of the benchmarks defined by name (e.g. `tpcds-3tb-delta`). Each benchmark specification is defined by the following: 
```

File: benchmarks/scripts/benchmarks.py (hunk 1 above)
Comment (tdas) [reply benchmarks/scripts/benchmarks.py:284]: Aargh. fixed it. thanks for testing this out!!

File: benchmarks/README.md (hunk 3 above)
Comment (tdas) [reply benchmarks/README.md:103]: done

File: benchmarks/README.md (hunk 2 above)
Comment (tdas) [reply benchmarks/README.md:37]: thats weird. .. i thought `#!/usr/bin/env python3` I have at the top of the run-benchmark.py should make it use python3.

**Pull request closed without merging**